### PR TITLE
Added LUT for associative ops and rfactor subgraph decomposition

### PR DESCRIPTION
**Pull request by psuriana**
Made rfactor more general with LUT of associative ops and added subgraph decomposition in associativity prover.

### CODE REVIEW

File: src/AssociativeOpsTable.h [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+
+// Represent an associative op with its identity. The op may be multi-dimensional,
+// e.g. complex multiplication. 'is_commutative' is set to true if the op is also
+// commutative in addition to being associative.
+struct AssociativePattern {
+    std::vector<Expr> ops;
+    std::vector<Expr> identities;
+    bool is_commutative;
+
+    AssociativePattern() : is_commutative(false) {}
```
Comment (abadams) [inline src/AssociativeOpsTable.h]: Need docstrings on all of these methods. Consider moving implementations into AssociativeOpsTable.cpp too - I don't think debugging print code belongs in a header (maybe that should go in IRPrinter with the rest of our operator<< overloads)

File: src/Associativity.cpp [hunk 2]
```diff
@@ -46,195 +59,385 @@ class ConvertSelfRef : public IRMutator {
...
+    // TODO(psuriana): there might be a better way to find all the transitive dependencies
+    bool change = true;
+    while (change) {
+        change = false;
+        for (size_t i = 0; i < dependencies.size(); ++i) {
+            for (size_t j = 0; j < dependencies.size(); ++j) {
+                if (i == j) {
+                    continue;
+                }
+                if (dependencies[i].find(j) != dependencies[i].end()) {
```
Comment (abadams) [inline src/Associativity.cpp]: dependencies[i].count(j)

File: src/Associativity.cpp [hunk 3]
```diff
@@ -46,195 +59,385 @@ class ConvertSelfRef : public IRMutator {
...
+                    for (const auto &idx : dependencies[j]) {
+                        if (dependencies[i].find(idx) == dependencies[i].end()) {
```
Comment (abadams) [inline src/Associativity.cpp]: It's a set, why not just insert it unconditionally?

Comment (abadams) [inline src/Associativity.cpp]: e.g. dependencies[i].insert(dependencies[j].begin(), dependencies[j].end())

File: src/Associativity.cpp [hunk 4]
```diff
@@ -46,195 +59,385 @@ class ConvertSelfRef : public IRMutator {
...
+        exprs[idx] = common_subexpression_elimination(exprs[idx]);
+        exprs[idx] = simplify(exprs[idx]);
+        exprs[idx] = solve_expression(exprs[idx], op_x_names[idx]).result; // Move 'x' to the left as possible
+        exprs[idx] = substitute_in_all_lets(exprs[idx]);
+    }
+    internal_assert((exprs.size() != 1) || all_independent) << "1D tuple should be all independent\n";
+
+    vector<set<int>> subgraphs;
+    if (!all_independent) {
+        debug(5) << "There is cross-dependencies. Need to prove associativity in bulk.\n";
```
Comment (abadams) [inline src/Associativity.cpp]: There is -> There are

File: test/performance/rfactor.cpp [hunk 5]
```diff
@@ -0,0 +1,423 @@
+#include "Halide.h"
```
Comment (abadams) [inline test/performance/rfactor.cpp]: It doesn't appear to be possible for this test to fail. I'd add some very conservative asserts on the improvement factors (e.g. check they don't make things substantially slower: improvement > 0.9) That should catch anything that breaks rfactor performance in some weird way.

File: src/Associativity.cpp (hunk 3 above)
Comment (psuriana) [inline src/Associativity.cpp]: I need to update the "change" flag so that I know when to break from the while loop

**Pull request merged**